conans/client/cmd/uploader.py

```diff
@@ -73,165 +69,65 @@ def add_prefs(self, prefs):
+        self._output.rewrite_line("Checking package integrity...")
 
-    def check_package(self, package, remote, policy):
-        pref = package.pref
-        assert (pref.revision is not None), "Cannot upload a package without PREV"
-        assert (pref.ref.revision is not None), "Cannot upload a package without RREV"
+        # short_paths = None is enough if there exist short_paths
+        pkg_layout = self._app.cache.pkg_layout(pref)
+        read_manifest, expected_manifest = pkg_layout.package_manifests()
 
-        try:
-            # TODO: It is a bit ugly, interface-wise to ask for revisions to check existence
-            server_revisions = self._remote_manager.get_package_revision_reference(pref, remote)
-            assert server_revisions
-        except NotFoundException:
-            package.upload = True
-            package.force = False
+        if read_manifest != expected_manifest:
+            self._output.writeln("")
+            diff = read_manifest.difference(expected_manifest)
+            for fname, (h1, h2) in diff.items():
+                self._output.warning("Mismatched checksum '%s' (manifest: %s, file: %s)"
+                                     % (fname, h1, h2))
+            raise ConanException("Cannot upload corrupted package '%s'" % str(pref))
         else:
-            if policy == UPLOAD_POLICY_FORCE:
-                self._output.info("{} already in server, forcing upload".format(pref.repr_notime()))
-                package.upload = True
-                package.force = True
-            else:
-                self._output.info("{} already in server, skipping upload".format(pref.repr_notime()))
-                package.upload = False
-                package.force = False
+            self._output.rewrite_line("Package integrity OK!")
+        self._output.writeln("")
 
 
-class _PackagePreparator:
-    def __init__(self, cache, remote_manager, hook_manager, loader):
-        self._cache = cache
-        self._remote_manager = remote_manager
+class PackagePreparator:
+    def __init__(self, app: ConanApp):
+        self._app = app
         self._output = ConanOutput()
-        self._hook_manager = hook_manager
-        self._loader = loader
 
-    def prepare(self, upload_data, remotes, integrity_check):
+    def prepare(self, upload_bundle, remote, force):
         self._output.info("Preparing artifacts to upload")
-        for recipe in upload_data.recipes:
-            layout = self._cache.ref_layout(recipe.ref)
+        for recipe in upload_bundle.recipes:
+            layout = self._app.cache.ref_layout(recipe.ref)
             conanfile_path = layout.conanfile()
-            conanfile = self._loader.load_basic(conanfile_path)
+            conanfile = self._app.loader.load_basic(conanfile_path)
             if recipe.upload:
-                self.prepare_recipe(recipe, conanfile, remotes)
+                self._check_upstream_recipe(recipe, remote, force)
```

Comment:
The ``check_upstream_recipe`` was explicitly separated from the ``Preparator`` to decouple the flow of preparing the packages, so it is possible to prepare a package, irrespective of what the server says or decide, in fact, without any server check at all.

I am fine with this, I think it is better to move forward and merge, but maybe we want to review the final ``UploadAPI``, I think having this separated is important.

Reply:
NP. If it is important I prefer to do it now. I'll move the check to the "upload_bundle" API method and later we could discuss if a new method for that check is convenient or not.

Reply:
I think if we want to offer this as an API, it needs to be separated:

- Select: Common to ``remove``, based on patterns, check the things from the cache  to do something (upload)
- Prepare: creates the zips (maybe check manifest integrity), of the packages that it is told to prepare, irrespective of server state.
- Filter-ExistingServer: Checks which revisions do exist in the server and then it is not necessary to upload or prepare
- Upload: Takes a prepared bundle and uploads it, as-is

Our built-in upload command is: Select + FilterExistingServer +Prepare + Upload. This way we avoid expensive zipping of things that don't need to be uploaded.

User custom commands can prepare things, irrespective of the server state.

Reply:
Yes, I was implementing that. I'll push in a min.